### PR TITLE
Add amp-ad support for Atomx

### PR DESCRIPTION
**Pull request by erikdubbelboer**
Atomx amp-ad support.

### CODE REVIEW

File: examples/ads.amp.html [hunk 1]
```diff
@@ -236,6 +237,12 @@ <h2 id="appnexus">AppNexus with JSON based configuration multi ad</h2>
       json='{"pageOpts":{"member": 958}, "adUnits": [{"disablePsa": true, "tagId": 6063968,"sizes": [300,250],"targetId": "apn_ad_1"}, {"tagId": 6063968,"sizes": [728,90],"targetId":"apn_ad_2"}]}'>
   </amp-ad>
 
+  <h2 id="atomx">Atomx</h2>
+  <amp-ad width=300 height=250
+      type="appnexus"
```
Comment (erwinmombay) [inline examples/ads.amp.html]: should the type be atomx?

Comment (erikdubbelboer) [inline examples/ads.amp.html]: Good catch, fixed.

**Comment by erikdubbelboer** [comment]
Fixed the one test that failed.

File: ads/atomx.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function atomx(global, data) {
+  const args = [];
+  args.push('size=' + data.width + 'x' + data.height);
+
+  // Validate empty fields
+  validateDataExists(data, ['id']);
```
Comment (lannka) [inline ads/atomx.js]: nitpick: please move validation code to the beginning of the method

**Comment by lannka** [comment]
@erikdubbelboer did you get it working locally? Run `gulp build --fortesting && gulp dist --fortesting` and visit `http://localhost:8000/examples/ads.amp.max.html#atomx` and `http://localhost:8000/examples/ads.amp.min.html#atomx` to check if it loads ads.

I got those errors in browser console:
```
p.js:1 GET http://p.ato.mx/placement?v=9&?size=300x250&id=1234&type=javascript&b=0&dom…es=1&flash=0&r=http%3A%2F%2Flocalhost%3A8000%2Fexamples%2Fads.amp.max.html (anonymous function) @ p.js:1(anonymous function) @ p.js:1
publishertag.js:1PubTag [DEBUG]:Publisher Tag loaded
pfx.min.js:1 PFXLOG: Can not get site.
pfx.min.js:1 PFXLOG: Blocked a frame with origin "http://ads.localhost:8000" from accessing a cross-origin frame.
pfx.min.js:1 PFXLOG: E-Service-010: 3df62288-9832-47d6-8797-0032bdeab1a4
```

File: ads/atomx.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  }
+  if (data.uv2) {
+    args.push('uv2=' + encodeURIComponent(data.uv2));
+  }
+  if (data.uv3) {
+    args.push('uv3=' + encodeURIComponent(data.uv3));
+  }
+  if (data.context) {
+    args.push('context=' + encodeURIComponent(data.context));
+  }
```
Comment (lannka) [inline ads/atomx.js]: I'm not strong about this, but you can probably make them in a loop since they all have the same pattern.

**Comment by erikdubbelboer** [comment]
@lannka I'm new to amphtml not quite sure how those tests are supposed to work? [examples/ads.amp.html is loading](https://github.com/ampproject/amphtml/blob/1463630ec720a75c363c07c6d710ddc89f1ee01d/examples/ads.amp.html#L41) https://cdn.ampproject.org/v0/amp-ad-0.1.js which obviously doesn't contain this change yet?

After running the gulp tasks I don't have a "ads.amp.max.html".

The only way I can get it to work is creating a .html file with:
```html
<!doctype html>
<html ⚡>
  <head>
    <meta charset="utf-8">
    <link rel="canonical" href="hello-world.html" >
    <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
    <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
    <script async custom-element="amp-ad" src="dist/v0/amp-ad-0.1.max.js"></script>
    <script async src="dist/amp.js"></script>
  </head>
  <body>
    <amp-ad width=300 height=250
      type="atomx"
      data-id="1234">
    </amp-ad>
  </body>
</html>
```
But this doesn't work either as it loads a two resources which don't exist:
https://3p.ampproject.net/1471590202566/f.js 
https://d-85807778709227184.ampproject.net/1471590202566/frame.html
I'm guessing also because this change hasn't been merged yet?

**Comment by lannka** [comment]
Ha, you just put `.max` or `.min` in the URL. magic will happen :-)

**Comment by erikdubbelboer** [comment]
It was unclear to me that I needed to use the gulp webserver to serve the files.

I fixed all the bugs and wrote the ifs to a loop.

**Comment by lannka** [comment]
@erikdubbelboer I tried your code but saw the error below:
```
p.js:1 GET http://p.ato.mx/placement?v=9&?size=300x250&id=1234&type=javascript&b=0&dom…es=1&flash=0&r=http%3A%2F%2Flocalhost%3A8000%2Fexamples%2Fads.amp.max.html 
```

Could you take a look?

Also, please sync to the latest master header, as there was a small refactoring you need to merge.

**Comment by erikdubbelboer** [comment]
@lannka I don't see any error there? Just that it's doing a GET?

I rebased against master.

**Comment by lannka** [comment]
@erikdubbelboer the GET was returning 400

**Comment by lannka** [comment]
Never mind, the problem seems gone.

LGTM

File: ads/atomx.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function atomx(global, data) {
+  const optionals = ['click', 'uv1', 'uv2', 'uv3', 'context'];
+  const args = [];
+
+  validateData(data, ['id'], optionals);
+
+  args.push('size=' + data.width + 'x' + data.height);
```
Comment (lannka) [inline ads/atomx.js]:
```
const args = [
  'size=' + data.width + 'x' + data.height,
  'id=' + encodeURIComponent(data.id),
];
```

**Comment by erikdubbelboer** [comment]
Fixed the minor comment.

**Pull request merged**